foyer/xml_writer.py

```diff
@@ -0,0 +1,281 @@
+                dihedral_force.set('id2', str(dihedral.atom2.idx))
+                dihedral_force.set('id3', str(dihedral.atom1.idx))
+                dihedral_force.set('id4', str(dihedral.atom4.idx))
+        else:
+            if unique:
+                if atypes[0] > atypes[-1]:
+                    atypes = atypes[::-1]
+            else:
+                dihedral_force.set('id1', str(dihedral.atom1.idx))
+                dihedral_force.set('id2', str(dihedral.atom2.idx))
+                dihedral_force.set('id3', str(dihedral.atom3.idx))
+                dihedral_force.set('id4', str(dihedral.atom4.idx))
+        for id in range(4):
+            dihedral_force.set('type{}'.format(id+1), atypes[id])
+        dihedral_force.set('periodicity1', str(dihedral.type.per))
+        dihedral_force.set('phase1',
+                           str(round(dihedral.type.phase * (np.pi / 180), 8)))
+        dihedral_force.set('k1', str(round(dihedral.type.phi_k * 4.184, 3)))
+
+def _write_rb_torsions(root, rb_torsions, unique):
+    rb_torsion_forces = ET.SubElement(root, 'RBTorsionForce')
+    for rb_torsion in rb_torsions:
+        rb_torsion_force = ET.SubElement(rb_torsion_forces, 'Proper')
+        atypes = [atype for atype in [rb_torsion.atom1.type,
+                                      rb_torsion.atom2.type,
+                                      rb_torsion.atom3.type,
+                                      rb_torsion.atom4.type]]
+        if unique:
+            if atypes[0] > atypes[-1]:
+                atypes = atypes[::-1]
+        else:
+            rb_torsion_force.set('id1', str(rb_torsion.atom1.idx))
+            rb_torsion_force.set('id2', str(rb_torsion.atom2.idx))
+            rb_torsion_force.set('id3', str(rb_torsion.atom3.idx))
+            rb_torsion_force.set('id4', str(rb_torsion.atom4.idx))
+        for id in range(4):
+            rb_torsion_force.set('type{}'.format(id+1), atypes[id])
+        for c_id in range(6):
+            rb_torsion_force.set('c{}'.format(c_id),
+                str(round(getattr(rb_torsion.type, 'c{}'.format(c_id)) * 4.184,
+                          4)))
+
+def _remove_duplicate_elements(root, unique):
+    sortby = {'AtomTypes': ['name'],
+              'HarmonicBondForce': ['type1', 'type2'],
+              'HarmonicAngleForce': ['type1', 'type2', 'type3'],
+              'PeriodicTorsionForce': ['type1', 'type2', 'type3', 'type4'],
+              'RBTorsionForce': ['type1', 'type2', 'type3', 'type4'],
+              'NonbondedForce': ['type']}
+    prev = None
+    for child in root:
+        if not unique and child.tag != 'AtomTypes':
+            continue
+        child[:] = sorted(child, key=lambda elem:
+            tuple(elem.get(id) for id in sortby[child.tag]))
+        elems_to_remove = []
+        for elem in child:
+            if _elements_equal(elem, prev):
+                elems_to_remove.append(elem)
+                continue
+            prev = elem
```

Comment:
Okay so if I'm interpreting this right, we're iterating through the XML elements and looking for match between the previous line and the current line. A match is defined as matching everything (like for a dihedral, type1/2/3/4, constants, ref values, periodicites etc). I think this works 99% of the time when each XML element refers to some unique bonded-tuple, and each XML element fully defines the bonded interaction for that bonded-tuple

Going back to the multiple-periodic torsion issue from above, we end up breaking this logic. In the ethane example, H-C-C-H could have multiple periodicities (let's say 2 periodicities). When this happens, the raw XML (still containing duplicates) will contain alternating XML elements with HCCH parameters for periodicity1, then HCCH parameters for periodicity2. Because we are just comparing previous and current XML elements, we effectively never compress the unique dihedrals. 

This seems to be a symptom/consequence of the earlier periodic torsion issues I outlined earlier - if the original issue gets addressed, I think this will get cured as well